File: ScaleformUI_Csharp/Menus/UIMenu/Items/UIMenuListItem.cs
using ScaleformUI.Elements;
using ScaleformUI.LobbyMenu;
using ScaleformUI.PauseMenu;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace ScaleformUI.Menu
{
    public class UIMenuListItem : UIMenuDynamicListItem, IListItem
    {
        protected internal int _index;
        protected internal List<dynamic> _items;
        public List<dynamic> IndexToValue;


        /// <summary>
        /// Triggered when the list is changed.
        /// </summary>
        public event ItemListEvent OnListChanged;

        /// <summary>        
        /// Triggered when a list item is selected.        
        /// </summary>        
        public event ItemListEvent OnListSelected;

        /// <summary>
        /// Returns the current selected index.
        /// </summary>
        public int Index
        {
            get { return Items.Count == 0 ? 0 : _index % Items.Count; }
            set
            {
                if (value < 0)
                    _index = Items.Count - 1;
                else if (value > Items.Count - 1)
                    _index = 0;
                else
                    _index = value;
                if (_items.Count > 0)
                    CurrentListItem = Items[_index].ToString();
                else
                    CurrentListItem = "";
            }
        }

        /// <summary>
        /// Returns the current selected index.
        /// </summary>
        public List<object> Items
        {
            get => _items;
            set
            {
                _items = new(value);
                if (_items.Count > 0)
                    CurrentListItem = Items[_index].ToString();
                else
                    CurrentListItem = "";
            }
        }

        /// <summary>
        /// List item, with left/right arrows.
        /// </summary>
        /// <param name="text">Item label.</param>
        /// <param name="items">List that contains your items.</param>
        /// <param name="index">Index in the list. If unsure user 0.</param>
        public UIMenuListItem(string text, List<dynamic> items, int index) : this(text, items, index, "")
        {
        }

        /// <summary>
        /// List item, with left/right arrows.
        /// </summary>
        /// <param name="text">Item label.</param>
        /// <param name="items">List that contains your items.</param>
        /// <param name="index">Index in the list. If unsure user 0.</param>
        /// <param name="description">Description for this item.</param>
        public UIMenuListItem(string text, List<dynamic> items, int index, string description) : this(text, items, index, description, SColor.HUD_Panel_light, SColor.White)
        {
        }

        private DynamicListItemChangeCallback _callback = async (sender, direction) =>
        {
            return await ((UIMenuListItem)sender).getIndex(direction);
        };

        public UIMenuListItem(string text, List<object> items, int index, string description, SColor mainColor, SColor higlightColor) : base(text, description, "")
        {
            _items = new(items);
            if (index > items.Count)
                Index = 0;
            else
                Index = index;
            Callback = _callback;
            if (items.Count > 0)
            {
                CurrentListItem = items[Index].ToString();
            }
        }

        private async Task<string> getIndex(ChangeDirection d)
        {
            if (d == ChangeDirection.Left)
            {
                _index--;
                if (_index < 0)
                    _index = Items.Count - 1;
            }
            else
            {
                _index++;
                if (_index >= Items.Count)
                    _index = 0;
            }
            return Items[_index].ToString();
        }

        /// <summary>
        /// Find an item in the list and return it's index.
        /// </summary>
        /// <param name="item">Item to search for.</param>
        /// <returns>Item index.</returns>
        [Obsolete("Use UIMenuListItem.Items.FindIndex(p => ReferenceEquals(p, item)) instead.")]
        public virtual int ItemToIndex(dynamic item)
        {
            return _items.FindIndex(p => ReferenceEquals(p, item));
        }

        /// <summary>
        /// Find an item by it's index and return the item.
        /// </summary>
        /// <param name="index">Item's index.</param>
        /// <returns>Item</returns>
        [Obsolete("Use UIMenuListItem.Items[Index] instead.")]
        public virtual dynamic IndexToItem(int index)
        {
            return _items[index];
        }

        /// <summary>
        /// Add a Panel to the UIMenuListItem
        /// </summary>
        /// <param name="panel"></param>
        public override void AddPanel(UIMenuPanel panel)
        {
            Panels.Add(panel);
            panel.SetParentItem(this);
        }

        internal virtual void ListChangedTrigger(int newindex)
        {
            OnListChanged?.Invoke(this, newindex);
        }

        internal virtual void ListSelectedTrigger(int newindex)
        {
            OnListSelected?.Invoke(this, newindex);
        }

        /// <summary>
        /// Change list dinamically
        /// </summary>
        /// <param name="list">The list that will replace the current one</param>
        /// <param name="index">Starting index</param>
        public void ChangeList(List<dynamic> list, int index)
        {
            _items = null;
            _items = new(list);
            Index = index;
            if(ParentColumn != null && ParentColumn.visible)
                ParentColumn.SendItemToScaleform(ParentColumn.Items.IndexOf(this), true);
            if (ParentColumn != null && ParentColumn.visible)
                ParentColumn.SendItemToScaleform(ParentColumn.Items.IndexOf(this), true);
        }

        public override void SetRightBadge(BadgeIcon badge)
        {
            throw new Exception("UIMenuListItem cannot have a right badge.");
        }

        public override void SetRightLabel(string text)
        {
            throw new Exception("UIMenuListItem cannot have a right label.");
        }

        [Obsolete("Use item.CurrentListItem instead.")]
        public string CurrentItem()
        {
            return CurrentListItem;
        }
    }
}

File: ScaleformUI_Lua/src/Menus/UIMenu/Items/UIMenuListItem.lua
UIMenuListItem = {}
UIMenuListItem.__index = UIMenuListItem
setmetatable(UIMenuListItem, { __index = UIMenuDynamicListItem })
UIMenuListItem.__call = function() return "UIMenuListItem" end

---@class UIMenuListItem : UIMenuDynamicListItem
---@field public AddPanel fun(self:UIMenuListItem, item:UIMenuStatisticsPanel|UIMenuPercentagePanel|UIMenuColorPanel|UIMenuGridPanel):nil

---New
---@param Text string
---@param Items table
---@param Index number
---@param Description string
function UIMenuListItem.New(Text, Items, Index, Description, color, highlightColor)
    if type(Items) ~= "table" then Items = {} end
    if Index == 0 then Index = 1 end

    local cb = function(sender, dir)
        if dir == "left" then
            self._Index = self._Index - 1
            if self._Index < 1 then
                self._Index = #self.Items
            end
        elseif dir == "right" then
            self._Index = self._Index + 1
            if self._Index > #self.Items then
                self._Index = 1
            end
        end
        return tostring(self.Items[self._Index])
    end

    local base = UIMenuDynamicListItem.New(Text or "", Description or "", "", cb, color or SColor.HUD_Panel_light, highlightColor or SColor.HUD_White)
    base.Items = Items
    base._Index = tonumber(Index) or 1
    base.OnListChanged = function(menu, item, newindex)
    end
    base.OnListSelected = function(menu, item, newindex)
    end
    if #base.Items > 0 then
        base:CurrentListItem(tostring(base.Items[base._Index]))
    else
        base:CurrentListItem("")
    end
    return setmetatable(base, UIMenuListItem)
end

---Index
---@param Index number
function UIMenuListItem:Index(Index)
    if tonumber(Index) then
        if Index > #self.Items then
            self._Index = 1
        elseif Index < 1 then
            self._Index = #self.Items
        else
            self._Index = Index
        end
        if #self.Items > 0 then
            self:CurrentListItem(tostring(self.Items[self._Index]))
        else
            self:CurrentListItem("")
        end
    else
        return self._Index
    end
end

---ItemToIndex
---@param Item table
function UIMenuListItem:ItemToIndex(Item)
    for i = 1, #self.Items do
        if type(Item) == type(self.Items[i]) and Item == self.Items[i] then
            return i
        elseif type(self.Items[i]) == "table" and (type(Item) == type(self.Items[i].Name) or type(Item) == type(self.Items[i].Value)) and (Item == self.Items[i].Name or Item == self.Items[i].Value) then
            return i
        end
    end
end

---IndexToItem
---@param Index number
function UIMenuListItem:IndexToItem(Index)
    if tonumber(Index) then
        if tonumber(Index) == 0 then Index = 1 end
        if self.Items[tonumber(Index)] then
            return self.Items[tonumber(Index)]
        end
    end
end

function UIMenuListItem:ChangeList(list, index)
    if index == nil or index < 1 or index > #list or self._Index > #list then index = 1 end
    if type(list) ~= "table" then return end
    self.Items = {}
    self.Items = list
    self:Index(index)

    if self.ParentMenu ~= nil and self.ParentMenu:Visible() then
        local it = IndexOf(self.ParentMenu.Items, self)
        self.ParentMenu:SendItemToScaleform(it, true)
    end
    if self.ParentColumn ~= nil then
        local it = IndexOf(self.ParentColumn.Items, self)
        self.ParentColumn:SendItemToScaleform(it, true)
    end
end